api/src/server.rs

```diff
@@ -202,30 +221,108 @@ impl<S: AccessModelStore> AccessModelStore for Arc<S> {
     }
 }
 
+struct EventInsert {
+    id: EventId,
+    data: Vec<u8>,
+    tx: tokio::sync::oneshot::Sender<Result<bool>>,
+}
+
+struct InsertTask {
+    _handle: tokio::task::JoinHandle<()>,
+    tx: tokio::sync::mpsc::Sender<EventInsert>,
+}
+
 #[derive(Clone)]
 pub struct Server<C, I, M> {
     peer_id: PeerId,
     network: Network,
     interest: I,
-    model: M,
+    model: Arc<M>,
+    // If we need to restart this ever, we'll need a mutex. For now we want to avoid locking the channel
+    // so we just keep track to gracefully shutdown, but if the task dies, the server is in a fatal error state.
+    insert_task: Arc<InsertTask>,
     marker: PhantomData<C>,
 }
 
 impl<C, I, M> Server<C, I, M>
 where
     I: AccessInterestStore,
-    M: AccessModelStore,
+    M: AccessModelStore + 'static,
 {
-    pub fn new(peer_id: PeerId, network: Network, interest: I, model: M) -> Self {
+    pub fn new(peer_id: PeerId, network: Network, interest: I, model: Arc<M>) -> Self {
+        let (tx, event_rx) = tokio::sync::mpsc::channel::<EventInsert>(1024);
+        let event_store = model.clone();
+
+        let handle = Self::start_insert_task(event_store, event_rx);
+        let insert_task = Arc::new(InsertTask {
+            _handle: handle,
+            tx,
+        });
         Server {
             peer_id,
             network,
             interest,
             model,
+            insert_task,
             marker: PhantomData,
         }
     }
 
+    fn start_insert_task(
+        event_store: Arc<M>,
+        mut event_rx: tokio::sync::mpsc::Receiver<EventInsert>,
+    ) -> tokio::task::JoinHandle<()> {
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_millis(FLUSH_INTERVAL_MS));
+            let mut events = vec![];
+            // could bias towards processing the queue of events over accepting more, but we'll
+            // rely on the channel depth for backpressure. the goal is to keep the queue close to empty
+            // without processing one at a time. when we stop parsing the carfile in the store
+            // i.e. validate before sending here and this is just an insert, we may want to process more at once.
+            loop {
+                tokio::select! {
+                    _ = async {}, if events.len() >= EVENT_INSERT_QUEUE_SIZE => {
```

Comment:
Can we just eliminate this match, and check after the select?

e.g. `select { tick, event }; if events.len() > process_events`

Reply:
lol yeah, good call. not sure why I didn't do that initially. Adding the `async {} if condition` felt... wrong.